### ISSUE TITLE
How do I perform cross validation with the Perceptron model?

### ISSUE THREAD

**Issue by benyarn** [mlpack/mlpack#3176]
### GitHub Intro ###

I am trying to implement my first mlpack model. I'm implementing KFoldCV with the perceptron model. I posted a stackoverflow question [here](https://stackoverflow.com/questions/71699577/how-to-perform-cross-validation-with-the-perceptron-model-in-mlpack), but I decided that the mlpack github would be a better place for this. I have copied and pasted the contents of the question below with minor modifications. 

### My StackOverflow Quesiton ###

I have been trying for several hours to perform cross-validation with the binary perceptron algorithm. I am following the example used with ```SoftmaxRegression``` given [here][1]. Here are the steps I have followed:

Step 1: (Works!) Get some random data.

```
// 100-point 6-dimensional random dataset.
arma::mat dataset = arma::randu<arma::mat>(6, 100);
// Random labels in the [0, 4] interval.
arma::Row<size_t> labels = arma::randi<arma::Row<size_t>>(100, arma::distr_param(0, 1));
```

Step 2: (Works?) Construct [KFoldCV object][2]. I have some options here and I'm not sure which is right. I put the following line in my code:

```
mlpack::cv::KFoldCV<mlpack::perceptron::Perceptron<>, mlpack::cv::Accuracy, arma::mat, arma::Row< size_t >> perceptron_cv(10, dataset, labels);
```

I thought this should match [this constructor][3], which has the comment

> This constructor can be used for regression algorithms and for binary classification algorithms.

in the documentation. However, the above line gives the error 

> The given MLAlgorithm requires the numClasses parameter

The version I have given below should call [this constructor][4].
```
mlpack::cv::KFoldCV<mlpack::perceptron::Perceptron<>, mlpack::cv::Accuracy, arma::mat, arma::Row< size_t >> perceptron_cv(10, dataset, labels, static_cast<size_t>(2));
```

The ```static_cast``` is used to overcome an ambiguous call error. The above line compiles. 

Step 3: (Doesn't work) The next step is to call the [evaluate member function][5]. The signature of this function is

```
double Evaluate	(	const MLAlgorithmArgs &... 	args	);
``` 

and the description for ```args``` from the documentation is

> args: Arguments for MLAlgorithm (in addition to the passed ones in the
> constructor).

From here, I did a deep dive into the source code to find out what ```args``` should be. It appears to me that ```args``` are passed as additional arguments to the constructor of ```Perceptron```. Here is the signature of the ```Perceptron``` constructor I want to call:

```
Perceptron (const MatType &data, const arma::Row< size_t > &labels, const size_t numClasses, const size_t maxIterations=1000);
```
It appears to me that no additional arguments need to be passed. So I call ```Evaluate```:

```
double perceptron_accuracy = perceptron_cv.Evaluate();
```

This gives the error

> The given MLAlgorithm is not constructible from the passed arguments

I can't figure out where to go from here.

  [1]: https://www.mlpack.org/doc/stable/doxygen/classmlpack_1_1cv_1_1KFoldCV.html#details
  [2]: https://www.mlpack.org/doc/stable/doxygen/classmlpack_1_1cv_1_1KFoldCV.html
  [3]: https://www.mlpack.org/doc/stable/doxygen/classmlpack_1_1cv_1_1KFoldCV.html#a7968536f97e0123e2af12bdbbae08d9d
  [4]: https://www.mlpack.org/doc/stable/doxygen/classmlpack_1_1cv_1_1KFoldCV.html#af0d5cf1c60beda485a47d8d5346d47fe
  [5]: https://www.mlpack.org/doc/stable/doxygen/classmlpack_1_1cv_1_1KFoldCV.html#a8e36d1ae2b8ec5f99d5d4a914a532c80

**Comment by benyarn** [comment]
This was meant to be tagged as a question, but I can't see how to add a "question" tag without closing the issue and reopening as an "Issue: Question."

**Comment by rcurtin** [comment]
I discovered the issue!  It turns out `Perceptron<>` does not have this constructor:

```
Perceptron(
    const MatType& data,
    const LabelsType& labels,
    const size_t numClasses,
    const WeightsType& weights);
```

Instead, it only has a constructor that does not take weights.  As a quick experiment, I added a `weights` parameter to the `Perceptron<>` constructor and it compiled fine.  However, it's late here, so it will be tomorrow until I manage to turn it into a working patch.
